api/sys_monitor.go

```diff
@@ -0,0 +1,62 @@
+package api
+
+import (
+	"bufio"
+	"context"
+)
+
+func (c *Sys) Monitor(logLevel string, stopCh chan struct{}) (chan string, error) {
+	r := c.c.NewRequest("GET", "/v1/sys/monitor")
+
+	if logLevel == "" {
+		r.Params.Add("log_level", "INFO")
+	} else {
+		r.Params.Add("log_level", logLevel)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+
+	logCh := make(chan string, 64)
+	go func() {
+		defer resp.Body.Close()
+		scanner := bufio.NewScanner(resp.Body)
+
+		for {
+			select {
+			case <-stopCh:
+				close(logCh)
+				cancelFunc()
+				return
+			case <-ctx.Done():
+				stopCh <- struct{}{}
+				close(logCh)
+				cancelFunc()
+				return
+			default:
+			}
+
+			if scanner.Scan() {
+				// An empty string signals to the caller that
+				// the scan is done, so make sure we only emit
+				// that when the scanner says it's done, not if
+				// we happen to ingest an empty line.
+				if text := scanner.Text(); text != "" {
+					logCh <- text
+				}
+			} else {
+				// If Scan() returns false, that means the context deadline was exceeded, so
+				// terminate this routine and start a new request.
+				stopCh <- struct{}{}
```

Comment:
We can't send safely to stopCh here because the caller may have closed it, causing this to panic.  Same goes for the send above in the ctx.Done case.  Let's revisit the context argument idea.

Reply:
I don't think we need to terminate and start a new request, based on a little prototyping I did with TestSysMonitorStreamingLogs rejiggered to use contexts.